lib/modules/datasource/julia-pkg-server/registry.ts

```diff
@@ -0,0 +1,246 @@
+import { Parse as tarParse } from 'tar';
+import { logger } from '../../../logger';
+import type { Http } from '../../../util/http';
+import { Result } from '../../../util/result';
+import { joinUrlParts } from '../../../util/url';
+import {
+  PKG_SERVER_REQUEST_HEADERS,
+  juliaPkgServerDatasourceId,
+} from './common';
+
+/**
+ * Generic container for aggregating and passing around the contents of files.
+ */
+interface FileContents<T extends Buffer[] | string> {
+  [path: string]: T;
+}
+
+/**
+ * Wraps all the information necessary to uniquely identify a registry at a
+ * particular point in time.
+ */
+interface JuliaPkgServerRegistry {
+  pkgServer: string;
+  state: string;
+  uuid: string;
+}
+
+/**
+ * URLs referencing a registry served by a PkgServer take the form:
+ * `<PkgServer host>/registry/<registry UUID>/<registry state>`.
+ *
+ * Although required when querying the state of a registry, the state in the
+ * URL is not mandatory when configuring a `registryUrl` passed into the
+ * datasource. Therefore that section is marked optional and should be ensured
+ * to be filled before further processing.
+ */
+const PKG_SERVER_REGISTRY_URL_FORMAT = new RegExp(
+  '^(?<pkgServer>.+)/registry/(?<uuid>\\p{Hex_Digit}{8}(-\\p{Hex_Digit}{4}){3}-\\p{Hex_Digit}{12})(/(?<state>\\p{Hex_Digit}{40}))?/?$',
+  'u',
+);
+
+/**
+ * The final part of the path of a (fully specified) registry URL references
+ * the tree SHA of the state of of the registry repository the package server
+ * is serving. Given the content-addressable nature of this SHA, it can be used
+ * as a caching mechanism.
```

Comment:
## Check input matches expectations

> I think this raises a good question as to whether we should be enforcing strict formatting on the state.

It's probably best to to check what Renovate "pulls in". So in general terms:

- Check input is valid
- Check input matches expected format
- Sanitize dangerous stuff
- Only then proceed with processing the input

## SHA state check as general utility/function to be used by other Renovate code?

Checking if the state is SHA-1 vs SHA-256 vs SHA-512 sounds like something that other Renovate code could re-use? Maybe that should be a new file/function in a suitable place?

## To check or not to check?

### I vote check

> Either we should probably check it everywhere, or we can forego the check altogether

I would vote check everywhere, but I'm not a maintainer. :smile:

### If we decide to not check, show good error message

> ... or we can forego the check altogether which will then just result in 404 errors when retrieving the registry ...

I guess that would look like this then?

1. Renovate tries to create PR for SHA-256 registry package.
2. Renovate shows generic 404 error.
3. User thinks Renovate is broken.

I would prefer something like this:

1. Renovate tries to create PR for SHA-256 registry package.
2. Renovate shows 404 error, and explains that the SHA-256 bit is not supported yet.
3. This probably prompts the user to ask the Renovate maintainers to add code to support the new hash.

## Your upside is my downside

> ... with the upside that it would not require changes to Renovate if the package server starts supporting different state formats.

I don't agree with this. If the package server supports a new format, we probably need to account for that in the Renovate code. So it's best if we state explicitly what's supported, what's not, and have a way to get nudged when things break.

## Maintainers make the decision here

I'll let the maintainers decide what to do with all this info above, and with your idea! :wink:

Reply:
can we simply allow sha256? so we're future prove when the package server becomes support for it? I think in near future more and more new repos will use sha256 by default.

Reply:
Thanks for the extensive input.

The state verification currently only applies to user-provided input. Given that a package server will directly report what URLs to use when asking it for the state of a registry, the URLs/states it provides are assumed to be correct (note that the package server hosts these URLs itself). If the package server turns out to provide wrong URLs that will indeed result in 404s.

Currently, no validation is done on those returned URLs (but I do have a note locally on whether this is something we'd want to do which I think is what you are suggesting). So I'm open to changing that.

> can we simply allow sha256?

I have added this in 1ffa7aee56078a9c57d0df50ddfa9af62c2a1c90.